src/proto/hybrid_vec.rs

```diff
@@ -0,0 +1,1832 @@
+use alloc::heap::{EMPTY, allocate, reallocate, deallocate};
+
+// Implemented on top of `core` to demonstrate the ability to substitute HybridVec for Vec
+use core::borrow::{BorrowFrom, ToOwned};
+use core::default::Default;
+use core::fmt;
+use core::hash::{self, Hash};
+use core::intrinsics::forget;
+use core::iter;
+use core::marker::ContravariantLifetime;
+use core::nonzero::NonZero;
+use core::mem;
+use core::num::{Int, UnsignedInt};
+use core::ops;
+use core::ptr;
+use core::raw::Slice as RawSlice;
+use core::usize;
+
+// Make this configurable?
+const ARRAY_SIZE: usize = 32;
+// The number of bytes a type has to be larger than to automatically be allocated on the heap.
+// This limits the on-stack size of `HybridVec` to a maximum of 
+// `ARRAY_SIZE * HEAP_BYTES_THRESHOLD`.
+const HEAP_BYTES_THRESHOLD: usize = 64;
+
+/// An implementation of `HybridVec` that uses stack-allocated storage below a certain length.
+/// A drop-in replacement for `Vec` from the stdlib, with a couple minor changes to the API and internal
+/// behavior.
+pub enum HybridVec<T> {
+    Stack {
+        array: [T; ARRAY_SIZE], 
+        len: usize,
+    },
+    Heap {
```

Comment:
Crap, I guess it could have been now that I think of it. That might have saved me a lot of time adapting the heap-management code, haha.

And then `IntoIter` and `Drain` could just contain the same types from `Vec` for the `Heap` variant. I'll work on that later tonight along with any other nits.